Runigram.java

```diff
@@ -60,48 +89,98 @@ private static void print(Color c) {
+		{
+			for(int j=0; j<image[0].length; j++)
+			{
+				System.out.print("(");
+				System.out.printf("%3s,",image[i][j].getRed());
+				System.out.printf("%3s,",image[i][j].getGreen());
+				System.out.printf("%3s",image[i][j].getBlue());
+				System.out.print(")  ");
+			}
+			System.out.println();
+		}
 	}
 	
 	/**
 	 * Returns an image which is the horizontally flipped version of the given image. 
 	 */
 	public static Color[][] flippedHorizontally(Color[][] image) {
-		//// Replace the following statement with your code
-		return null;
+		Color[][] secondImage=new Color[image.length][image[0].length];
+		for(int i=0; i<secondImage.length; i++)
+		{
+			for(int n=0, j=image[0].length-1 ;n<image[0].length && j>=0 ;j--, n++)
+			{
+				secondImage[i][n]=image[i][j]; //image[i][j]
+			}
+		}
+
+		return secondImage;
 	}
 	
 	/**
 	 * Returns an image which is the vertically flipped version of the given image. 
 	 */
 	public static Color[][] flippedVertically(Color[][] image){
-		//// Replace the following statement with your code
-		return null;
+		Color[][] secondImage=new Color[image.length][image[0].length];
+		for(int j=0; j<secondImage[0].length; j++)
+		{
+			for(int i=0, n=image.length-1 ;n>=0 && i<image.length ;i++, n--)
+			{
+				secondImage[i][j]=image[n][j]; 
+			}
+		}
+		return secondImage;
 	}
 	
 	// Computes the luminance of the RGB values of the given pixel, using the formula 
 	// lum = 0.299 * r + 0.587 * g + 0.114 * b, and returns a Color object consisting
 	// the three values r = lum, g = lum, b = lum.
 	public static Color luminance(Color pixel) {
-		//// Replace the following statement with your code
-		return null;
+		int r=pixel.getRed();
+		int g=pixel.getGreen();
+		int b=pixel.getBlue();
+		int lum = (int)(0.299 * r + 0.587 * g + 0.114 * b);
+		Color newPixel=new Color(lum,lum,lum);
+		return newPixel;
 	}
 	
 	/**
 	 * Returns an image which is the grayscaled version of the given image.
 	 */
 	public static Color[][] grayScaled(Color[][] image) {
-		//// Replace the following statement with your code
-		return null;
+		Color[][] newImage=new Color[image.length][image[0].length];
+		for(int i=0; i<newImage.length; i++)
+		{
+			for(int j=0; j<newImage[0].length; j++)
+			{
+				newImage[i][j]=luminance(image[i][j]);
+			}
+		}
+		return newImage;
 	}	
 	
 	/**
 	 * Returns an image which is the scaled version of the given image. 
 	 * The image is scaled (resized) to have the given width and height.
 	 */
 	public static Color[][] scaled(Color[][] image, int width, int height) {
-		//// Replace the following statement with your code
-		return null;
+		Color[][] newImage=new Color[height][width];
+		// int h=(image.length/height);
+		// int w=(image[0].length/width);
+		int originalHeight=image.length;
+		int originalWidth=image[0].length;
+		for(int i=0; i<height; i++)
+		{
+			for(int j=0; j<width; j++)
+			{
+				int h= i*originalHeight/height;
+				int w= j*originalWidth/width;

```

Comment:
These calculations are not dependant on loop iterators, try to avoid calculate same value multiple times. 
```suggestion
		double heightScale = originalHeight/height;
		double widthScale = originalWidth/width;
		for(int i=0; i<height; i++)
		{
			int h= I*heightScale;
			for(int j=0; j<width; j++)
			{
				int w= j*widthScale;

```